Mathlib/GroupTheory/GroupAction/Jordan.lean

```diff
@@ -0,0 +1,673 @@
+  have _ : Fintype α := Fintype.ofFinite α
+  have _ : Fintype G := Fintype.ofFinite G
+  apply Subgroup.eq_top_of_card_eq
+  simp only [Nat.card_eq_fintype_card]
+  apply le_antisymm
+  · apply Fintype.card_subtype_le
+  · rw [Fintype.card_equiv (Equiv.cast rfl)]
+    trans (2 : ℕ).factorial
+    · rw [← Nat.card_eq_fintype_card]
+      exact Nat.factorial_le hα
+    · simpa [Nat.factorial_two, ← Fintype.one_lt_card_iff_nontrivial] using hG
+
+theorem nontrivial_on_equiv_perm_two [Finite α] {K : Type*} [Group K] [MulAction K α]
+    (hα : Nat.card α = 2) (hK : fixedPoints K α ≠ .univ) :
+    IsMultiplyPretransitive K α 2 := by
+  classical
+  let φ := MulAction.toPermHom K α
+  let f : α →ₑ[φ] α :=
+    { toFun := id
+      map_smul' := fun _ _ ↦ rfl }
+  have hf : Function.Bijective f := Function.bijective_id
+  suffices Function.Surjective φ by
+    unfold IsMultiplyPretransitive
+    rw [IsPretransitive.of_embedding_congr this hf (n := Fin 2), ← hα]
+    apply Perm.isMultiplyPretransitive
+  rw [← MonoidHom.range_eq_top]
+  apply Subgroup.eq_top_of_card_eq
+  apply le_antisymm (card_le_card_group φ.range)
+  simp only [Nat.card_perm, hα, Nat.factorial_two]
+  by_contra H
+  simp only [not_le, Nat.lt_succ, Finite.card_le_one_iff_subsingleton] at H
+  apply hK
+  apply Set.eq_univ_of_univ_subset
+  intro a _ g
+  suffices φ g = φ 1 by
+    conv_rhs => rw [← one_smul K a]
+    simp only [← toPerm_apply, ← toPermHom_apply K α g]
+    exact congrFun (congrArg DFunLike.coe this) a
+  simpa [← Subtype.coe_inj] using H.elim ⟨_, ⟨g, rfl⟩⟩ ⟨_, ⟨1, rfl⟩⟩
+
+variable [Fintype α] [DecidableEq α]
+
+theorem isPretransitive_of_cycle {g : Equiv.Perm α}
+    (hg : g ∈ G) (hgc : g.IsCycle) :
+    IsPretransitive (fixingSubgroup G ((↑g.support : Set α)ᶜ))
+      (SubMulAction.ofFixingSubgroup G ((↑g.support : Set α)ᶜ)) := by
+  obtain ⟨a, _, hgc⟩ := hgc
+  have hs : ∀ x : α, g • x ≠ x ↔
+    x ∈ SubMulAction.ofFixingSubgroup G ((↑g.support : Set α)ᶜ) := by
+    intro x
+    simp [SubMulAction.mem_ofFixingSubgroup_iff]
+  suffices ∀ x ∈ SubMulAction.ofFixingSubgroup G ((↑g.support : Set α)ᶜ),
+      ∃ k : fixingSubgroup G ((↑g.support : Set α)ᶜ), x = k • a by
+    rw [isPretransitive_iff]
+    rintro ⟨x, hx⟩ ⟨y, hy⟩
+    obtain ⟨k, hk⟩ := this x hx
+    obtain ⟨k', hk'⟩ := this y hy
+    use k' * k⁻¹
+    rw [← SetLike.coe_eq_coe]
+    simp only [SetLike.mk_smul_mk]
+    rw [hk, hk', smul_smul, inv_mul_cancel_right]
+  intro x hx
+  have hg' : (⟨g, hg⟩ : ↥G) ∈ fixingSubgroup G ((↑g.support : Set α)ᶜ) := by
+    simp_rw [mem_fixingSubgroup_iff G]
+    intro y hy
+    simpa only [Set.mem_compl_iff, Finset.mem_coe, Equiv.Perm.notMem_support] using hy
+  let g' : fixingSubgroup (↥G) ((↑g.support : Set α)ᶜ) := ⟨(⟨g, hg⟩ : ↥G), hg'⟩
+  obtain ⟨i, hi⟩ := hgc ((hs x).mpr hx)
+  exact ⟨g' ^ i, hi.symm⟩
+
+/-- A primitive permutation group that contains a swap is the full permutation group (Jordan) -/
+theorem jordan_swap (hG : IsPreprimitive G α) (g : Equiv.Perm α)
```

Comment:
This lemma and the following will need more descriptive names.

Reply:
Done. Tell me if you feel they don't work.